packages/@aws-cdk/assertions/lib/match.ts

```diff
@@ -1,13 +1,242 @@
 import { ABSENT } from './vendored/assert';
 
 /**
- * Partial and special matching during template assertions
+ * Denotes a failure when matching patterns to targets.
  */
-export class Match {
+export interface MatchFailure {
+  /**
+   * The relative path at which the failure occurred.
+   */
+  readonly path: string[];
+  /**
+   * Description of the match failure.
+   */
+  readonly message: string;
+}
+
+/**
+ * Partial and special matching during template assertions.
+ */
+export abstract class Match {
   /**
    * Use this matcher in the place of a field's value, if the field must not be present.
    */
   public static absentProperty(): string {
     return ABSENT;
   }
+
+  /**
+   * Matches the specified pattern with the array found in the same relative path of the target.
+   * The set of elements (or matchers) must be in the same order as would be found.
+   * @param pattern the pattern to match
+   * @param options options to configure the matcher
+   */
+  public static arrayWith(pattern: any[], options: ArrayMatchOptions = {}): Match {
+    return new ArrayMatch(pattern, options);
+  }
+
+  /**
+   * Matches the specified pattern to an object found in the same relative path of the target.
+   * The keys and their values (or matchers) must be present in the target but the target can be a superset.
+   * @param pattern the pattern to match
+   * @param options options to configure the matcher
+   */
+  public static objectLike(pattern: {[key: string]: any}, options: ObjectMatchOptions = {}): Match {
+    return new ObjectMatch(pattern, options);
+  }
+
+  /**
+   * Check whether the provided object is a subtype of the `Match` class
+   */
+  public static isMatcher(x: any): x is Match {
```

Comment:
I feel like we should unify on "match" vs "matcher"

Reply:
`isMatch()` feels like "is it a match?" but willing to be persuaded the other way.

Reply:
`isMatchType`?